### DISCUSSION TITLE
Frontend geospatial drawing and selection

### DISCUSSION THREAD

**Discussion by pratishta** [NYCPlanning/ae-zoning-map-poc#66] [General]
Related to https://github.com/NYCPlanning/ae-zoning-map-poc/issues/63. Branch[ 63-using-nebula-gl](https://github.com/NYCPlanning/ae-zoning-map-poc/tree/63-using-nebula-gl)

### Initial findings
Documentation is somewhat outdated. Installing nebula [as described](https://nebula.gl/docs/get-started/installing) causes errors and is ineffective. I've had to manually install the modules I ended up with the following:
```
npm install "@nebula.gl/layers"
npm install "@nebula.gl/edit-modes"
```

[Running the example](https://github.com/uber/nebula.gl/blob/master/README.md#running-the-example) provided in the nebula README was helpful. 

### Current issues
There are numerous type errors. For example, when creating a new editable layer: 
```
const editableLayer = new (EditableGeoJsonLayer as any)({
    id: 'geojson',
    data: shapes,
    mode,
    onEdit: (obj: Object) => {
      console.log(obj);
    },
    selectedFeatureIndexes,
  });
```
we get 
<img width="596" alt="image" src="https://github.com/NYCPlanning/ae-zoning-map-poc/assets/7049943/b29a6233-d2d3-4296-9446-b6f896baede1">
<img width="596" alt="image" src="https://github.com/NYCPlanning/ae-zoning-map-poc/assets/7049943/e225d7f2-f6aa-41e6-9152-a5e1f48dfca3">

I went with using this [solution](https://github.com/uber/nebula.gl/issues/568#issuecomment-836324975) from a similar issue that was raised in the nebula repo. Having type ambiguity defeats the purpose of a lot of the tools we use, and this issue is from over two years ago without an update. 

I'll start building out the features for drawing and downloading and continue documenting here.

**Comment by pratishta** [comment]
Brief detour looking into [terra-draw](https://github.com/JamesLMilner/terra-draw)

**Comment by pratishta** [comment]
Okay there's a working version with terra draw on [this branch](https://github.com/NYCPlanning/ae-zoning-map-poc/tree/use-terra-draw)
<img width="1913" alt="image" src="https://github.com/NYCPlanning/ae-zoning-map-poc/assets/7049943/f6595f7f-f02a-465e-9761-a3f08bfc344a">

Next steps would be to figure out integration with the drawn polygons on a new deck layer instead of directly on the basemap itself.

**Comment by pratishta** [comment]
Continuing to look into terra-draw because it's agnostic to mapping libraries, has extensive documentation, and an active contributor. (It's maybe even too new?) But for now, it still capable of drawing and measuring (https://terradraw.io/) and with a bit more digging, I believe we can write our own method for downloading the GeoJSON and hook it up with geospatial filtering to get tax lots within a polygon. 

One thing that comes up is whether or not we want to persist the drawing GeoJSON so we can bring it back after the state changes. My impression is probably no? I assume we're not storing any user interaction with zoning currently but I know there was some interest from the user interviews

**Reply by TangoYankee** [reply]
> One thing that comes up is whether or not we want to persist the drawing GeoJSON so we can bring it back after the state changes. My impression is probably no? I assume we're not storing any user interaction with zoning currently but I know there was some interest from the user interviews

I want to split hairs on the idea of "state" changing. There is in-memory state that lives as long as the application- until the user refreshes the page, navigates away, closes the browser, or otherwise clears application memory. We get his behavior built-in with each drawing tool. But then, there is external state that lives on after the application memory is wiped. This is the feature that I believe users have been asking for. And, there are ways to achieve it without complex infrastructure. Take [geojson.io](https://geojson.io/) as an example. They persist geojson to a local storage variable called "recover", which the application updates as you interact with it. Then when you return to the site later, you can pick up where you left off. (fun fact, [wordle](https://www.nytimes.com/games/wordle/index.html) also does this) 

Keeping localstorage updated and recovering the data on each load are still problems we would need to solve. However, they exist independent of the drawing tools. Whatever drawing tool we decide on, they should all be compatible with our eventual persistence approach.

**Reply by TangoYankee** [reply]
On investigating Zustand, it has tooling [for persisting its state](https://github.com/pmndrs/zustand?tab=readme-ov-file#persist-middleware) to localstorage or something similar

**Comment by pratishta** [comment]
Added a section for the different drawing modes. We can now have polygons, lines, points, and circles and the ability to select and move those shapes around. The drawn shapes are stored in [The Store](https://github.com/JamesLMilner/terra-draw/blob/main/guides/2.STORE.md) as a GeoJSON. 

<img width="1696" alt="image" src="https://github.com/NYCPlanning/ae-zoning-map-poc/assets/7049943/188cec75-922c-43db-a5ac-22409b9f84ed">

Currently terra-draw doesn't support rotation of polygons but every point clicked by the user is editable in the 'select' mode. 
<img width="1198" alt="image" src="https://github.com/NYCPlanning/ae-zoning-map-poc/assets/7049943/ea10d182-7594-45fc-8183-50b6cfcf8349">
<img width="1198" alt="image" src="https://github.com/NYCPlanning/ae-zoning-map-poc/assets/7049943/4dcfbb83-ce7a-4071-8ed7-d3bae72dd09a">

Next steps to me look like:
* Display spatial data of the figures (like the side panel on https://terradraw.io/)
* Download the store GeoJSON
* Pass a polygon feature to the geospatial filter 🤞🙏

**Reply by pratishta** [reply]
Also: there's a lot of ugly react happening on this branch. Beware

**Reply by TangoYankee** [reply]
If memory serves, Deck.gl and TerraDraw are boxing each other out to render layers and cpature events. Deck gl renders layers [on top of the basemap](https://deck.gl/docs/get-started/using-with-map). TerraDraw acts directly on on the basemap. In turn, the Deck gl layers hide the TerraDraw layers and capture user events before they can get to the basemap and subsequently TerraDraw.

**Reply by pratishta** [reply]
By default, yes we have the base map that's overlaid with the deck layers. By using the [interleaved option](https://github.com/visgl/deck.gl/blob/e0cefb0b2bf5efc21d62fff04a23f6d6b5f90b33/docs/developer-guide/base-maps/using-with-maplibre.md#interleaved) though, it's possible to capture the TerraDraw events on the base map.

```
const layers = [taxLotsLayer, zoningDistrictsLayer];
  function DeckGLOverlay(props: MapboxOverlayProps) {
    const overlay = useControl(() => new DeckOverlay(props));
    overlay.setProps(props);
    return null;
  }

return (
       <Map
            ref={(ref) => setMapRefObj(ref)}
            initialViewState={viewState}
            style={{ width: "100vw", height: "100vh" }}
            mapStyle="https://raw.githubusercontent.com/NYCPlanning/equity-tool/main/src/data/basemap.json"
            // disable the default attribution
            attributionControl={false}
          >
          <DeckGLOverlay layers={layers} interleaved={true} />
        </Map> 
)
```
But this now allows events from the deck layers (if they're on) and base map to both be picked up so I'll need to have coordinate selection of TerraDraw objects and tax lots / zoning districts.

**Comment by TangoYankee** [comment]
I played around with creating a drawing tool directly in a Deck.gl layer with help from turf.js Code is on: [bespoke-drawing-store](https://github.com/NYCPlanning/ae-remix-template/tree/bespoke-drawing-store)

For functionality, there are modes specific to drawing points and linestrings (...with the intent to support squares, right triangles, and/or free-hand polygons, etc). There are also multiple ways to end/cancel drawing. 1) Clicking again on the selector buttons 2) Pressing "Enter" 3) clicking again on the last drawn point or 4) Reaching the maximum number of allowed points in a linestring.

For code structure, it relies on the reducer pattern outlined in this [client state discussion](https://github.com/NYCPlanning/ae-zoning-map-poc/discussions/26#discussioncomment-7661992). It also uses Immer to isolate and control changes to the store, improving rendering predictability and performance. ~~I chose the reducer and context hooks over Zustand because Zustand seems to obfuscate much of the control over state updates, hampering the ability to reason about and control state changes.~~ Upon revisiting Zustand for simpler use cases, there is literally an [Immer + Zustand section](https://github.com/pmndrs/zustand?tab=readme-ov-file#sick-of-reducers-and-changing-nested-states-use-immer) in Zustand's readme.

**Reply by TangoYankee** [reply]
(It's a little messy because it's transitioning away from the first pass at an implementation found on https://github.com/NYCPlanning/ae-remix-template/tree/bespoke-drawing)

**Reply by TangoYankee** [reply]
This branch has point, line, and rectangle modes in the zustand store. https://github.com/NYCPlanning/ae-zoning-map-poc/tree/point-line-rectangle

**Comment by pratishta** [comment]
Okay, updated use of terra-draw still on this branch https://github.com/NYCPlanning/ae-zoning-map-poc/tree/use-terra-draw

To use this locally you need to update line 49 of file `decode-params-pipes.ts` in the zoning-api from `const values = value.split(",");` to `const values = Array.isArray(value) ? value : value.split(",");`

A heads up, I didn't prioritize UI for this spike and mostly just wanted to functionality of terra-draw to see if it can do what we need.

How to use:
1. Click on 'draw' button in Interaction Modes (You are only able to draw when this button is active)
2. Draw polygon on map
3. Click 'select' button
4. Click on polygon
-> Should populate with first 20 tax lots and shape info on the upper right hand corner
<img width="1690" alt="image" src="https://github.com/NYCPlanning/ae-zoning-map-poc/assets/7049943/c0c91639-a78e-438b-bad8-3ac313e0542d">

For linestring:
1. Click 'linestring'
2. Draw a line
3. Click 'select' button
4. Click on line

-> Should populate with line info
<img width="1690" alt="image" src="https://github.com/NYCPlanning/ae-zoning-map-poc/assets/7049943/00fa2eba-bedc-4a66-8db3-a3f93ce71209">

When 'draw' is not active, and a zoning or tax lot layer is on, the usual information shows up without the selected polygon tax lots or shape information. 
<img width="1690" alt="image" src="https://github.com/NYCPlanning/ae-zoning-map-poc/assets/7049943/2a56125e-0afa-4037-8ba3-01224269d1ce">

When the layers are off, the selected tax lots and shape information shows up.
<img width="1690" alt="image" src="https://github.com/NYCPlanning/ae-zoning-map-poc/assets/7049943/81d3924c-502a-4e18-a502-5ae9399adc19">

This is pretty buggy because I haven't yet had time to properly use terra's built in [event management](https://github.com/JamesLMilner/terra-draw/blob/main/guides/6.EVENTS.md) to manage states. Additionally how the client state discussion comes into play here, especially re: rendering performance, I'm not sure. As far as I can tell, the biggest change is using the [interleaved method](https://github.com/NYCPlanning/ae-zoning-map-poc/discussions/66#discussioncomment-8988747) of displaying the base map and Deck.gl layers. I haven't gotten to being able to download the polygons but we're able to access all created features with `terraDraw.getSnapshot()` and should able to [store and restore data](https://github.com/JamesLMilner/terra-draw/blob/main/guides/2.STORE.md#restoring-data).

In terms of the amount of work for what we need, I've come back around to preferring terra-draw.

**Reply by pratishta** [reply]
Some more caveats / bugs:
- Clicking on a created polygon's lines creates more points on those lines, causing the polygon sent to the api to have more than 5 points (which breaks findTaxLots)
- Creating a second polygon limits you to 3 points (triangle) and immediately starts another. My best guess is that I'm not finishing a drawing properly / not specifying some restrictions for polygon creation

**Reply by TangoYankee** [reply]
> To use this locally you need to update line 49 of file decode-params-pipes.ts in the zoning-api from const values = value.split(","); to const values = Array.isArray(value) ? value : value.split(",");

This code is available on PR nycplanning/ae-zoning-api#194

**Comment by TylerMatteo** [comment]
@pratishta @TangoYankee Really great work on this research! It looks like we have two viable alternatives to Nebula. We can put a pin in this for now. For now, I'd lean slightly towards Tim's custom Deck layer implementation, if only because Terradraw is an alpha library and relies on integrating with Maplibre and Deck's interleaved basemap approach. Using Maplibre and interleaved layers isn't necessarily bad, but it does add some complexity compared to keeping things more contained in Deck-land. The big pro I see of Terradraw is that it provides a nicer UI and event API out of the box. Once we have "real world" use cases to implement, the right choice should become clearer.

**Comment by TylerMatteo** [comment]
@pratishta @TangoYankee Just fyi - in a recent Open Visualization bi-weekly meeting, the Vis.gl maintainers gave an update on the future of Nebula.gl (timestamped youtube link [here](https://youtu.be/tFd-zuLIHw4?si=BTaI89QEBafi6qkC&t=347)). The tldr is that they moved it into the [deck.gl community repo ](https://github.com/visgl/deck.gl-community)and have been working on [updates](https://github.com/visgl/deck.gl-community/issues/38). I think the approaches we have here are still great but it's nice to know that nebula isn't necessarily totally dead.
